Firestore/core/src/firebase/firestore/remote/grpc_stream.mm

```diff
@@ -0,0 +1,368 @@
+  request.labels = [serializer_ encodedListenRequestLabelsForQueryData:query];
+
+  return ToByteBuffer([request data]);
+}
+
+grpc::ByteBuffer ObjcBridge::ToByteBuffer(FSTTargetID target_id) const {
+  GCFSListenRequest* request = [GCFSListenRequest message];
+  request.database = [serializer_ encodedDatabaseID];
+  request.removeTarget = target_id;
+
+  return ToByteBuffer([request data]);
+}
+
+grpc::ByteBuffer ObjcBridge::ToByteBuffer(NSData* data) const {
+  const grpc::Slice slice{[data bytes], [data length]};
+  return grpc::ByteBuffer{&slice, 1};
+}
+
+FSTWatchChange* ObjcBridge::GetWatchChange(GCFSListenResponse* proto) {
+  return [serializer_ decodedWatchChange:proto];
+}
+
+SnapshotVersion ObjcBridge::GetSnapshotVersion(GCFSListenResponse* proto) {
+  return [serializer_ versionFromListenResponse:proto];
+}
+
+NSData* ObjcBridge::ToNsData(const grpc::ByteBuffer& buffer) const {
+  std::vector<grpc::Slice> slices;
+  buffer.Dump(&slices);  // TODO: check return value
+  if (slices.size() == 1) {
+    return [NSData dataWithBytes:slices.front().begin()
+                          length:slices.front().size()];
+  } else {
+    // FIXME
+    return [NSData dataWithBytes:slices.front().begin()
+                          length:slices.front().size()];
+  }
+}
+
+void BufferedWriter::Enqueue(grpc::ByteBuffer&& bytes) {
+  buffer_.push_back(std::move(bytes));
+  TryWrite();
+}
+
+void BufferedWriter::TryWrite() {
+  if (!is_started_) {
+    return;
+  }
+  if (buffer_.empty()) {
+    return;
+  }
+  // From the docs:
+  // Only one write may be outstanding at any given time. This means that
+  /// after calling Write, one must wait to receive \a tag from the completion
+  /// queue BEFORE calling Write again.
+  if (has_pending_write_) {
+    return;
+  }
+
+  has_pending_write_ = true;
+  call_->Write(buffer_.back(), &kWriteTag);
+  buffer_.pop_back();
+}
+
+void BufferedWriter::OnSuccessfulWrite() {
+  has_pending_write_ = false;
+  TryWrite();
+}
+
+}  // namespace internal
+
+using firebase::firestore::auth::CredentialsProvider;
+using firebase::firestore::auth::Token;
+using firebase::firestore::core::DatabaseInfo;
+using firebase::firestore::model::DatabaseId;
+// using firebase::firestore::model::SnapshotVersion;
+
+WatchStream::WatchStream(util::AsyncQueue* const async_queue,
+                         const DatabaseInfo& database_info,
+                         CredentialsProvider* const credentials_provider,
+                         FSTSerializerBeta* serializer)
+    : database_info_{&database_info},
+      firestore_queue_{async_queue},
+      credentials_provider_{credentials_provider},
+      stub_{CreateStub()},
+      dedicated_executor_{CreateExecutor()},
+      objc_bridge_{serializer} {
+  dedicated_executor_->Execute([this] {
+    PollGrpcQueue();
+  });
+}
+
+WatchStream::~WatchStream() {
+  grpc_queue_.Shutdown();
+  dedicated_executor_->ExecuteBlocking([]{});
+}
+
+void WatchStream::PollGrpcQueue() {
```

Comment:
If I understand the gRPC world correctly generally you set up one completion queue per channel, so it seems like this executor should exist at the level of FSTDatastore.

That in turn means that the tag values you use need to encode both which RPC is doing the work and the callback to invoke.

The easiest path forward would be to do something like what we did with the executor, where you new up an Operation object and submit that as the tag. The Operation object can then have a `std::shared_ptr` to the Stream.

Alternatively, the Datastore object could have a set of pending operations that it owns and when we shut a stream down we sever the connection between the pending operation and the stream.

In any case this arrangement is going to be hard to make work in the presence of multiple streams on a channel.